pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/GitHubBug207.txt

```diff
@@ -4,13 +4,13 @@
       +- ClassBody[@Empty = false, @Size = 1]
          +- MethodDeclaration[@Abstract = false, @Arity = 1, @EffectiveVisibility = Visibility.V_PRIVATE, @Final = false, @Name = "resourceHttpMessageWriter", @Overridden = false, @Static = true, @Varargs = false, @Visibility = Visibility.V_PRIVATE, @Void = false]
             +- ModifierList[@EffectiveModifiers = (JModifier.PRIVATE, JModifier.STATIC), @ExplicitModifiers = (JModifier.PRIVATE, JModifier.STATIC)]
-            +- ClassType[@FullyQualified = false, @SimpleName = "HttpMessageWriter"]
+            +- ClassType[@FullyQualified = false, @PackageQualifier = null, @SimpleName = "HttpMessageWriter"]
             |  +- TypeArguments[@Diamond = false, @Empty = false, @Size = 1]
-            |     +- ClassType[@FullyQualified = false, @SimpleName = "Resource"]
+            |     +- ClassType[@FullyQualified = false, @PackageQualifier = null, @SimpleName = "Resource"]
             +- FormalParameters[@Empty = false, @Size = 1]
             |  +- FormalParameter[@EffectiveVisibility = Visibility.V_LOCAL, @Final = false, @Varargs = false, @Visibility = Visibility.V_LOCAL]
             |     +- ModifierList[@EffectiveModifiers = (), @ExplicitModifiers = ()]
-            |     +- ClassType[@FullyQualified = true, @SimpleName = "Context"]
+            |     +- ClassType[@FullyQualified = true, @PackageQualifier = "BodyInserter", @SimpleName = "Context"]
```

Comment:
The actual code line is this: https://github.com/pmd/pmd/blob/33fd84d521517d91fbc020cadecc84f7e21329f4/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/GitHubBug207.java#L2

We don't have a correct auxclasspath here (I guess), so we can't be sure of anything. But intuitively I would expect that this would be `@FullyQualified = false` and `@PackageQualifier = null` (mostly, because of naming conventions - package names are usually all lowercase).

In real-world code snippets, there would be probably an import statement, which might help to figure this out.

Not sure, if we should just ignore this file (GitHubBug207) as a "not so useful" test case...

Reply:
> I would expect that this would be @FullyQualified = false and @PackageQualifier = null

I thought it would behave that way too... Will investigate.

The recovery behavior in case of incomplete auxclasspath doesn't use naming conventions, and I think we shouldn't change that for now. In this case I think it should be preserving an AmbiguousName in the tree...

Reply:
So I found that the recovery behavior for this kind of type interprets "BodyInserter" as a package name. This is the simplest decision we can make at this point. If we were to reinterpret the name using naming conventions to separate the package parts from the class parts, we would complicate the code a lot. Ultimately the AST structure is wrong, but a warning is emitted, and the user should fix their auxclasspath. I think we can dismiss this test.

Reply:
Yes, let's keep it simple. Without a complete auxclasspath (and in this example, without proper `import` statements - the code never compiles), we can't do anything better.